domain/secret/state/state.go

```diff
@@ -357,30 +415,277 @@ func (st State) createSecret(
+		checkExists = st.checkApplicationSecretLabelExists(existing[0].Owner.ID)
+	case coresecrets.UnitOwner:
+		// Query selects the unit uuid as owner id.
+		checkExists = st.checkUnitSecretLabelExists(existing[0].Owner.ID)
+	default:
+		// Should never happen.
+		return errors.Errorf("unexpected secret owner kind %q", kind)
+	}
+
+	if secret.Label != nil && *secret.Label != "" {
+		if err := checkExists(ctx, tx, *secret.Label); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	now := time.Now().UTC()
+	dbSecret := secretMetadata{
+		ID:             dbSecrets[0].ID,
+		Version:        dbSecrets[0].Version,
+		Description:    dbSecrets[0].Description,
+		AutoPrune:      dbSecrets[0].AutoPrune,
+		RotatePolicyID: int(domainsecret.MarshallRotatePolicy(&existing[0].RotatePolicy)),
+		UpdateTime:     now,
+	}
+	dbSecret.UpdateTime = now
+	updateSecretMetadataFromParams(secret, &dbSecret)
+	if err := st.upsertSecret(ctx, tx, dbSecret); err != nil {
+		return errors.Annotatef(err, "updating secret %q", uri)
+	}
+
+	if secret.Label != nil {
+		if err := st.upsertSecretLabel(ctx, tx, existing[0].URI, *secret.Label, existing[0].Owner); err != nil {
+			return errors.Annotatef(err, "updating label for secret %q", uri)
+		}
+	}
+
+	// Will secret rotate? If not, delete next rotation row.
+	if secret.RotatePolicy != nil && *secret.RotatePolicy == domainsecret.RotateNever {
+		deleteNextRotate := "DELETE FROM secret_rotation WHERE secret_id=$M.id"
+		deleteNextRotateStmt, err := st.Prepare(deleteNextRotate, sqlair.M{})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = tx.Query(ctx, deleteNextRotateStmt, secretIDParam).Run()
+		if err != nil {
+			return errors.Annotatef(err, "deleting next rotate record for secret %q", uri)
+		}
+	}
+
+	if len(secret.Data) == 0 && secret.ValueRef == nil {
+		return nil
+	}
+
+	revisionUUID, err := uuid.NewUUID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	nextRevision := existing[0].LatestRevision + 1
+	dbRevision := secretRevision{
+		ID:         revisionUUID.String(),
+		SecretID:   uri.ID,
+		Revision:   nextRevision,
+		CreateTime: now,
+		UpdateTime: now,
+	}
+	if err := st.upsertSecretRevision(ctx, tx, dbRevision, secret.ExpireTime); err != nil {
+		return errors.Annotatef(err, "inserting revision for secret %q", uri)
+	}
+
+	if len(secret.Data) > 0 {
+		if err := st.updateSecretContent(ctx, tx, dbRevision.ID, secret.Data); err != nil {
+			return errors.Annotatef(err, "updating content for secret %q", uri)
+		}
+	}
+
+	if secret.ValueRef != nil {
+		if err := st.upsertSecretValueRef(ctx, tx, dbRevision.ID, secret.ValueRef); err != nil {
+			return errors.Annotatef(err, "updating backend value reference for secret %q", uri)
+		}
+	}
+
+	if err := st.markObsoleteRevisions(ctx, tx, uri, nextRevision); err != nil {
+		return errors.Annotatef(err, "marking obsolete revisions for secret %q", uri)
+	}
+
+	if dbSecret.AutoPrune {
+		if err := st.pruneUnusedRevisions(ctx, tx, uri); err != nil {
```

Comment:
Should the obsolete revisions be pruned after the agents receive obsolete events and delete the content from the provider?

Reply:
This was just a port of the existing logic. Note that is is for user secrets, not charm owned secrets so there's no agents involved. If we think the existing logic needs to be changed, we can do that but need to recall why it was done this way originally.